pkg/kube/secretcontroller/secretcontroller.go

```diff
@@ -0,0 +1,269 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secretcontroller
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+
+	//"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	mcLabel    = "istio/multiCluster"
+	maxRetries = 5
+)
+
+// Required for unit test override.
+var loadKubeConfig = clientcmd.Load
+
+// AddClusterFunc prototype for the add secret callback function.
+type AddClusterFunc func(clientset kubernetes.Interface, dataKey string) error
```

Comment:
I would rename it to 'addSecret' - and also pass the Secret value ( to avoid another lookup ).
Long term we may use a lighter k8s library without a cache ( in particular if we add secret controller to k8s-pilot-agent)

Reply:
Changed to add/removeSecretCallback. Didn't add secret data in add callback. Seems like a bit of a security issue passing that data around. We can add in future if we find that a consumer of this library really needs that info.